scripts/check_backend_associated_test_file.py

```diff
@@ -0,0 +1,131 @@
+    'scripts/linters/test_files/invalid_request.py',
+    'scripts/linters/test_files/invalid_tabs.py',
+    'scripts/linters/test_files/invalid_todo.py',
+    'scripts/linters/test_files/invalid_urlopen.py',
+    'scripts/linters/test_files/valid.py',
+    'scripts/linters/test_files/valid_py_ignore_pragma.py',
+    'core/tests/build_sources/extensions/CodeRepl.py',
+    'core/tests/build_sources/extensions/DragAndDropSortInput.py',
+    'core/tests/data/failing_tests.py',
+    'core/tests/data/image_constants.py',
+    'core/tests/data/unicode_and_str_handler.py',
+    'proto_files/text_classifier_pb2.py',
+    'proto_files/training_job_response_payload_pb2.py',
+
+    # The files below are to be tested.
+    # GSoC'22: Improve line and branch coverage for the backend and frontend.
+    'scripts/run_custom_eslint_tests.py',
+    'scripts/run_backend_tests.py',
+    'scripts/run_lighthouse_tests.py',
+    'scripts/run_portserver.py',
+    'scripts/run_presubmit_checks.py',
+    'scripts/run_tests.py',
+    'scripts/start.py',
+    'setup.py',
+    'scripts/check_backend_associated_test_file.py',
+    'scripts/third_party_size_check.py',
+    'scripts/linters/warranted_angular_security_bypasses.py',
+    'core/feconf.py',
+    'core/domain/cron_services.py',
+    'core/platform/datastore/cloud_datastore_services.py',
+    'core/platform/transactions/cloud_transaction_services.py',
+]
+
+
+def main() -> None:
+    """Finds the non-empty backend files that lack an associated test file."""
+
+    all_backend_files = []
+    for (root, _, files) in os.walk(TOPMOST_LEVEL_PATH):
+        for file in files:
+            full_path = os.path.join(root, file)
+            _, file_extension = os.path.splitext(full_path)
+            if (
+                    file_extension == '.py' and
+                    '/third_party/' not in full_path and
+                    '/node_modules/' not in full_path
+            ):
+                all_backend_files.append(full_path)
+
+    all_backend_files.sort()
+    files_without_test = []
+    for file in all_backend_files:
+        if not file.endswith('_test.py'):
+            # Replace last occurance of '.py' with '_test.py' to get the
+            # associated test file.
+            test_file_name = '%s_test.py' % file[:-3]
+            if test_file_name not in all_backend_files:
+                files_without_test.append(file)
+
+    non_empty_files = []
+    for file in files_without_test:
+        line_count = 0
+        with open(file, 'r', encoding='utf8') as f:
+            line_count = len(f.readlines())
+        if line_count > 0:
+            non_empty_files.append(file[2:])
+
+    errors = ''
+    for file in non_empty_files:
+        if file not in FILES_WITHOUT_ASSOCIATED_TEST_FILES:
+            errors += (
+                '\033[1m{}\033[0m does not have an associated test file.'
+                ' Make sure it\'s fully tested by its associated test'
+                ' file.\n'.format(file))
```

Comment:
I think we can simplify this error message to: "{some file} needs an associated backend test file." I also like that this revised message communicates that this script doesn't check whether the associated test file provides full coverage. We're only checking that the test file exists.

Reply:
Updated it now.